src/inflators/model.tsx

```diff
@@ -1,30 +1,82 @@
+import { mapMaterials } from "../utils/material-utils";
+import { EntityID } from "../utils/networking-types";
 
 function camelCase(s: string) {
   return s.replace(/-(\w)/g, (_, m) => m.toUpperCase());
 }
 
 export type ModelParams = { model: Object3D };
 
+// These components are all handled in some special way, not through inflators
+const ignoredComponents = ["visible", "frustum", "frustrum", "shadow", "networked"];
+
+function inflateComponents(
+  world: HubsWorld,
+  eid: number,
+  components: { [componentName: string]: any },
+  idx2eid: Map<number, EntityID>
+) {
+  Object.keys(components).forEach(name => {
+    const inflatorName = camelCase(name);
+    if (ignoredComponents.includes(inflatorName)) return;
+    if (!gltfInflatorExists(inflatorName)) {
+      console.warn(`Failed to inflate unknown component called ${inflatorName}`, components[name]);
+      return;
+    }
+
+    const props = components[name];
+    Object.keys(components[name]).forEach(propName => {
+      const value = props[propName];
+      const linkType = value?.__mhc_link_type;
+      if (linkType) {
+        if (linkType !== "node") {
+          throw new Error("Non node link types should be resolved before inflateModel is called");
+        }
+        const existingEid = idx2eid.get(value.index);
+        if (existingEid) {
+          props[propName] = existingEid;
+        } else {
+          props[propName] = addEntity(world);
+          idx2eid.set(value.index, props[propName]);
+        }
+      }
+    });
+    gltfInflators[inflatorName](world, eid, props);
+  });
+}
+
 export function inflateModel(world: HubsWorld, rootEid: number, { model }: ModelParams) {
   const swap: [old: Object3D, replacement: Object3D][] = [];
+  const idx2eid = new Map<number, number>();
   model.traverse(obj => {
-    const components = obj.userData.gltfExtensions?.MOZ_hubs_components || {};
-    const eid = obj === model ? rootEid : addEntity(world);
-    Object.keys(components).forEach(name => {
-      const inflatorName = camelCase(name);
-      if (inflatorName === "visible" || inflatorName === "frustum" || inflatorName === "shadow") {
-        // These are handled below
-      } else if (!gltfInflatorExists(inflatorName)) {
-        console.warn(`Failed to inflate unknown component called ${inflatorName}`, components[name]);
-      } else {
-        gltfInflators[inflatorName](world, eid, components[name]);
-      }
+    const gltfIndex: number | undefined = obj.userData.gltfIndex;
+
+    let eid: number;
+    if (obj === model) {
+      eid = rootEid;
+    } else if (gltfIndex !== undefined && idx2eid.has(gltfIndex)) {
+      eid = idx2eid.get(gltfIndex)!;
+    } else {
+      eid = addEntity(world);
+    }
+
+    if (gltfIndex !== undefined) idx2eid.set(gltfIndex, eid);
+
+    const components = obj.userData.gltfExtensions?.MOZ_hubs_components;
+    if (components) inflateComponents(world, eid, components, idx2eid);
+
+    mapMaterials(obj, function (mat: Material) {
```

Comment:
Should this check if the object was replaced by an inflator?
`mapMaterials( world.eid2obj.get(eid) || obj, `

Reply:
The materials in the GLTF file certainly would not be on an object created by an inflator. If `obj` is a `Mesh` (only thing that can have materials from the GLTFLoader) and swapped, that's a bug, and we throw an exception below.